activemq-broker/src/main/java/org/apache/activemq/replica/ReplicaSequenceStorage.java

```diff
@@ -0,0 +1,149 @@
+import org.apache.activemq.util.IdGenerator;
+import org.apache.activemq.util.LongSequenceGenerator;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static java.util.Objects.requireNonNull;
+
+public class ReplicaSequenceStorage {
+
+    private final Logger logger = LoggerFactory.getLogger(ReplicaSequenceStorage.class);
+
+    static final String SEQUENCE_NAME_PROPERTY = "SequenceName";
+    private final LongSequenceGenerator eventMessageIdGenerator = new LongSequenceGenerator();
+    private final ProducerId replicationProducerId = new ProducerId();
+    private final Broker broker;
+    private final ConnectionContext connectionContext;
+    private final ReplicaInternalMessageProducer replicaInternalMessageProducer;
+    private final String sequenceName;
+    private final ReplicaReplicationQueueSupplier queueProvider;
+
+    private Queue sequenceQueue;
+    private PrefetchSubscription subscription;
+
+    public ReplicaSequenceStorage(Broker broker, ConnectionContext connectionContext, ReplicaReplicationQueueSupplier queueProvider,
+                                  ReplicaInternalMessageProducer replicaInternalMessageProducer, String sequenceName) {
+        this.broker = requireNonNull(broker);
+        this.connectionContext = connectionContext;
+        this.replicaInternalMessageProducer = replicaInternalMessageProducer;
+        this.sequenceName = requireNonNull(sequenceName);
+        this.queueProvider = queueProvider;
+
+        replicationProducerId.setConnectionId(new IdGenerator().generateId());
+    }
+
+    public String initialize() throws Exception {
+        sequenceQueue = broker.getDestinations(queueProvider.getSequenceQueue()).stream().findFirst()
+            .map(DestinationExtractor::extractQueue).orElseThrow();
+
+        String selector = String.format("%s LIKE '%s'", SEQUENCE_NAME_PROPERTY, sequenceName);
+
+        ConnectionId connectionId = new ConnectionId(new IdGenerator("ReplicationPlugin.ReplicaSequenceStorage").generateId());
+        SessionId sessionId = new SessionId(connectionId, new LongSequenceGenerator().getNextSequenceId());
+        ConsumerId consumerId = new ConsumerId(sessionId, new LongSequenceGenerator().getNextSequenceId());
+        ConsumerInfo consumerInfo = new ConsumerInfo();
+        consumerInfo.setConsumerId(consumerId);
+        consumerInfo.setPrefetchSize(10);
+        consumerInfo.setDestination(queueProvider.getSequenceQueue());
+        consumerInfo.setSelector(selector);
+        subscription = (PrefetchSubscription) broker.addConsumer(connectionContext, consumerInfo);
+
+        List<ActiveMQTextMessage> allMessages = new ArrayList<>();
+        for (MessageId messageId : sequenceQueue.getAllMessageIds()) {
+            ActiveMQTextMessage message = getMessageByMessageId(messageId);
+            if (message.getStringProperty(SEQUENCE_NAME_PROPERTY).equals(sequenceName)) {
```

Comment:
You already have the selector checking this right?

Reply:
yes, but it's a different piece of code. here we are making sure that there is only one message for the sequence. and if there are more than one we need to delete all except the latest.
it's just a safety mechanism.